### PR TITLE
Print Memory Requirements Before Real Allocation

### PR DESCRIPTION
**Pull request by pengwa**
**Description**: Display activation memory needed for Large Models

When we tried to understand the gaps on memory usage for large models (that cannot fit on one single GPU), we always need to know how much memory planned for initializers (e.g. training parameters) and activations (e.g. intermediate tensors). 

Initializers memory can be easily got from either the initializer planner (that happens before any CUDA memory allocation), or inferred from the parameter counts based on our knowledge on memory consumption.

For activations, we don't have such information.  This PR did this by just simply trying to generate a memory pattern earlier (for example before the initializer CUDA memory allocation) at the "best efforts", then we print it out, even though it will potentially fail later. The "best efforts" means, we try, but not necessary succeed. In the case where training graph have dim_param values, we will not have a successful memory pattern generated here, this is expected.

#### Sample outputs as below:

> OrtMemoryInfo:[name:CudaPinned id:0 OrtMemType:-1 OrtAllocatorType:1 Device:[DeviceType:0 MemoryType:1 DeviceId:0]]Activation Peak: Allocated memory for activations, size: 1048576
> OrtMemoryInfo:[name:Cpu id:0 OrtMemType:0 OrtAllocatorType:1 Device:[DeviceType:0 MemoryType:0 DeviceId:0]]Activation Peak: Allocated memory for activations, size: 513806912
> OrtMemoryInfo:[name:Cuda id:3 OrtMemType:0 OrtAllocatorType:1 Device:[DeviceType:1 MemoryType:0 DeviceId:3]]Activation Peak: Allocated memory for activations, size: 1395579008

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

File: onnxruntime/core/framework/finalize_session_state.cc [hunk 1]
```diff
@@ -82,6 +82,23 @@ Status FinalizeSessionState(SessionState& session_state,
       ITensorAllocator::Create(session_state.GetEnableMemoryPattern(), *exec_plan_ptr, session_state,
                                session_state.GetMutableWeightsBuffers()));
 
+#ifdef ENABLE_TRAINING
+  if (session_state.GetEnableMemoryPattern()) {
```
Comment (edgchen1) [inline onnxruntime/core/framework/finalize_session_state.cc:86]: consider also adding "&& logging at INFO is enabled"

File: onnxruntime/core/framework/finalize_session_state.cc [hunk 2]
```diff
@@ -193,7 +210,8 @@ common::Status SaveInitializedTensors(const Env& env, const std::basic_string<PA
   //2. allocate weight buffer on different locations
   // planned_initializers_memory_size_in_byte is not actual physical size.
   // It's the virtual size computed by planner.
-  std::unordered_map<std::string, size_t> planned_initializers_memory_sizes_in_byte;
+  std::unordered_map<std::string, size_t>
+      planned_initializers_memory_sizes_in_byte;
```
Comment (edgchen1) [inline onnxruntime/core/framework/finalize_session_state.cc:214]: don't need to start a new line here

File: onnxruntime/core/framework/session_state.cc [hunk 3]
```diff
@@ -345,17 +345,21 @@ Status SessionState::GeneratePatternGroupCache(const std::vector<std::reference_
...
-  auto& node_index_info = GetNodeIndexInfo();
   for (auto& node_plan : exe_plan->execution_plan) {
-    int node_index = node_index_info.GetNodeOffset(node_plan.node_index);
     auto* node = graph_viewer_->GetNode(node_plan.node_index);
-    int output_start = node_index + static_cast<int>(node->InputDefs().size()) + static_cast<int>(node->ImplicitInputDefs().size());
     //allocate output
     for (int i = 0, end = static_cast<int>(node->OutputDefs().size()); i < end; ++i) {
-      const auto ml_value_idx = node_index_info.GetMLValueIndex(output_start + i);
+      int ml_value_idx;
+      this->ort_value_name_idx_map_.GetIdx(node->OutputDefs()[i]->Name(), ml_value_idx);
```
Comment (edgchen1) [inline onnxruntime/core/framework/session_state.cc:362]: nit: don't need "this->"

Comment (edgchen1) [reply onnxruntime/core/framework/session_state.cc:362]: check return value of GetIdx()

Comment (edgchen1) [reply onnxruntime/core/framework/session_state.cc:362]: actually checking the return value may not be needed here - comparing with kInvalidEntry should be good

Comment (pengwa) [reply onnxruntime/core/framework/session_state.cc:362]: i saw other logic called GetIdx did check the status, let me do it. :) Thanks for catching this,.

File: onnxruntime/core/framework/finalize_session_state.cc (hunk 2 above)
Comment (pengwa) [reply onnxruntime/core/framework/finalize_session_state.cc:214]: good catch!

File: onnxruntime/core/framework/finalize_session_state.cc (hunk 1 above)
Comment (pengwa) [reply onnxruntime/core/framework/finalize_session_state.cc:86]: Fair

**Comment by pengwa** [comment]
@SherlockNoMad  I saw you changed the memory planner too, let me know whether we still need this change or not. :)

**Comment by snnn** [comment]
Please reopen it if it is still needed.

**Pull request closed without merging**